### PR TITLE
1352-oneContributionPerRole

### PR DESCRIPTION
**Pull request by TobiasNx**
Upbranched from #1371 and #1376
See for issue #1352

It seems that here are some problems with the replace_all at the end of the contribution.fix somehow this does not replace all instances. Perhaps this has something to do with the complex do list construction?

Could you help @fsteeg or @blackwinter

### CODE REVIEW

File: src/test/resources/alma-fix/(DE-605)HT019246898.json [hunk 1]
```diff
@@ -15,10 +15,22 @@
...
+        "id" : "http://id.loc.gov/vocabulary/relators/aut",
+        "label" : "Autor/in"
+      }
+    },
+    "type" : [ "Contribution" ]
   }, {
     "agent" : {
       "gndIdentifier" : "(DE-588)1038489679",
-      "id" : "https://d-nb.info/gnd/1038489679",
+      "id" : "(DE-588)1038489679",
```
Comment (TobiasNx) [inline src/test/resources/alma-fix/(DE-605)HT019246898.json:33]: The id should be changed but here id has not been.
The fix function that should change it should be:
https://github.com/hbz/lobid-resources/blob/6f9419cc5a2263465cb62c4eba6c695739247310/src/main/resources/alma/fix/contribution.fix#L174-175

Could you have a look what causes the error @blackwinter or @fsteeg

**Comment by blackwinter** [comment]
It has to do with `id` being copied from the same source field to multiple target objects.

A minimal reproduction for `HT019246898` would be:

```perl
set_array("contribution[]")
do list(path: "7001 ", "var": "$i")
  copy_field("$i.0.1", "contribution[].$append.id")
  copy_field("$i.0.1", "contribution[].$append.id")
end
replace_all("contribution[].*.id", "^\\(DE-588\\)(.*$)", "https://d-nb.info/gnd/$1")
retain("contribution[]")
```

Which results in:

```json
{
  "contribution" : [ {
    "id" : "(DE-588)1069204668"
  }, {
    "id" : "https://d-nb.info/gnd/1069204668"
  }, {
    "id" : "(DE-588)1038489679"
  }, {
    "id" : "https://d-nb.info/gnd/1038489679"
  } ]
}
```

I can investigate more tomorrow.

**Comment by blackwinter** [comment]
It works with `paste()` instead of `copy_field`:

```perl
set_array("contribution[]")
do list(path: "7001 ", "var": "$i")
  paste("contribution[].$append.id", "$i.0.1")
  paste("contribution[].$append.id", "$i.0.1")
end
replace_all("contribution[].*.id", "^\\(DE-588\\)(.*$)", "https://d-nb.info/gnd/$1")
retain("contribution[]")
```

```json
{
  "contribution" : [ {
    "id" : "https://d-nb.info/gnd/1069204668"
  }, {
    "id" : "https://d-nb.info/gnd/1069204668"
  }, {
    "id" : "https://d-nb.info/gnd/1038489679"
  }, {
    "id" : "https://d-nb.info/gnd/1038489679"
  } ]
}
```

**Comment by TobiasNx** [comment]
> It works with `paste()` instead of `copy_field()`:
> 
> ```perl
> set_array("contribution[]")
> do list(path: "7001 ", "var": "$i")
>   paste("contribution[].$append.id", "$i.0.1")
>   paste("contribution[].$append.id", "$i.0.1")
> end
> replace_all("contribution[].*.id", "^\\(DE-588\\)(.*$)", "https://d-nb.info/gnd/$1")
> retain("contribution[]")
> ```
> 
> ```json
> {
>   "contribution" : [ {
>     "id" : "https://d-nb.info/gnd/1069204668"
>   }, {
>     "id" : "https://d-nb.info/gnd/1069204668"
>   }, {
>     "id" : "https://d-nb.info/gnd/1038489679"
>   }, {
>     "id" : "https://d-nb.info/gnd/1038489679"
>   } ]
> }
> ```

Thanks. For the `copy_field`-behaviour we should still open another FIX-issue right?

**Comment by blackwinter** [comment]
Yes, that's certainly a bug.

**Comment by TobiasNx** [comment]
This PR can be closed it is solved by #1379

**Pull request closed without merging**